### PR TITLE
[Infineon PSoC6] Fix for Door lock and Software Diagnostics Cluster

### PR DESCRIPTION
**Pull request by woody-apple**
#### Problem
Fix for DRLK2.2 , DRLK-2.3 and DRLK-2.5
Fix for ThreadMetrics read in Software Diagnostics Cluster

#### Change overview
Update LockManager to support Lock/Unlock state properly based on Pin configuration
Add Support for ThreadMetrics in Software Diagnostics Cluster
Updated POOL Buffer to 48 to fix Packet Empty intermittent error

#### Testing
Manually tested DRLK2.2 , DRLK-2.3 and DRLK-2.5 using Chip-tool
Verified thread metrics read using chip-tool
Heap usage/free verified after pool buffer increase

### CODE REVIEW

File: src/platform/P6/DiagnosticDataProviderImpl.cpp [hunk 1]
```diff
@@ -544,6 +544,59 @@ CHIP_ERROR DiagnosticDataProviderImpl::WiFiCounters(WiFiStatsCountType type, uin
...
+    if (taskStatusArray != NULL)
+    {
+        /* Generate raw status information about each task. */
+        arraySize = uxTaskGetSystemState(taskStatusArray, arraySize, &dummy);
+        /* For each populated position in the taskStatusArray array,
+           format the raw data as human readable ASCII data. */
+
+        for (x = 0; x < arraySize; x++)
+        {
+            ThreadMetrics * thread = (ThreadMetrics *) pvPortMalloc(sizeof(ThreadMetrics));
```
Comment (andy31415) [inline src/platform/P6/DiagnosticDataProviderImpl.cpp:566]: Should we `VerifyOrDie(thread != nullptr)` ?

File: src/platform/P6/DiagnosticDataProviderImpl.cpp [hunk 2]
```diff
@@ -544,6 +544,59 @@ CHIP_ERROR DiagnosticDataProviderImpl::WiFiCounters(WiFiStatsCountType type, uin
...
+            thread->Next = head;
+            head         = thread;
+        }
+
+        *threadMetricsOut = head;
+        /* The array is no longer needed, free the memory it consumes. */
+        vPortFree(taskStatusArray);
+    }
+
+    return CHIP_NO_ERROR;
```
Comment (andy31415) [inline src/platform/P6/DiagnosticDataProviderImpl.cpp:587]: if taskStatusArray is NULL, why do we still return CHIP_NO_ERROR? should we at least log a memory allocation error?

File: src/platform/P6/DiagnosticDataProviderImpl.cpp [hunk 3]
```diff
@@ -544,6 +544,59 @@ CHIP_ERROR DiagnosticDataProviderImpl::WiFiCounters(WiFiStatsCountType type, uin
...
+        /* Generate raw status information about each task. */
+        arraySize = uxTaskGetSystemState(taskStatusArray, arraySize, &dummy);
+        /* For each populated position in the taskStatusArray array,
+           format the raw data as human readable ASCII data. */
+
+        for (x = 0; x < arraySize; x++)
+        {
+            ThreadMetrics * thread = (ThreadMetrics *) pvPortMalloc(sizeof(ThreadMetrics));
+
+            strncpy(thread->NameBuf, taskStatusArray[x].pcTaskName, kMaxThreadNameLength - 1);
```
Comment (andy31415) [inline src/platform/P6/DiagnosticDataProviderImpl.cpp:568]: nit: `chip::Platform::CopyString` from CHIPMemString.h does this safe copy with ensuring null terminator.

```cpp
chip::Platform::CopyString(thread->NameBuf, taskStatusArray[x].pcTaskName);
```

File: src/platform/P6/DiagnosticDataProviderImpl.cpp (hunk 1 above)
Comment (woody-apple) [reply src/platform/P6/DiagnosticDataProviderImpl.cpp:566]: This is a cherry pick.

File: src/platform/P6/DiagnosticDataProviderImpl.cpp (hunk 2 above)
Comment (woody-apple) [reply src/platform/P6/DiagnosticDataProviderImpl.cpp:587]: This is a cherry pick.

File: src/platform/P6/DiagnosticDataProviderImpl.cpp (hunk 3 above)
Comment (woody-apple) [reply src/platform/P6/DiagnosticDataProviderImpl.cpp:568]: This is a cherry pick.

**Pull request merged**